MacawTests/SceneSerialization.swift

```diff
@@ -0,0 +1,529 @@
+//
+//  SceneSerialization.swift
+//  MacawTests
+//
+//  Created by Alisa Mylnikova on 23/04/2018.
+//  Copyright © 2018 Exyte. All rights reserved.
+//
+
+import UIKit
+import Macaw
+
+protocol Initializable {
+    init()
+}
+
+extension Double : Initializable {}
+extension Bool : Initializable {}
+extension Int : Initializable {}
+
+func parse<T: Initializable>(_ from: Any?) -> T {
+    return from as? T ?? T()
+}
+
+
+internal protocol Serializable {
+    func toDictionary() -> [String:Any]
+}
+
+extension Node {
+    
+    func baseToDictionary() -> [String:Any] {
+        var result = ["place": place.toString(), "opaque": String(describing: opaque), "opacity": String(describing: opacity)] as [String : Any]
+        if let clip = clip as? Serializable {
+            result["clip"] = clip.toDictionary()
+        }
+        return result
+    }
+    
+    func fromDictionary(dictionary: [String:Any]) {
+        place = Transform(string: dictionary["place"] as? String)
+        opaque = Bool(dictionary["opaque"] as? String ?? "") ?? true
+        opacity = Double(dictionary["opacity"] as? String ?? "") ?? 0
+        if let locusDict = dictionary["clip"] as? [String:Any] {
+            clip = Locus.instantiate(dictionary: locusDict)
+        }
+    }
+    
+    static func instantiate(dictionary: [String:Any]) -> Node? {
+        guard let nodeType = dictionary["node"] as? String else {
+            fatalError("No node type specified")
+        }
+        
+        if nodeType == "Shape" {
+            return Shape(dictionary: dictionary)
+        }
+        
+        if nodeType == "Text" {
+            return Text(dictionary: dictionary)
+        }
+        
+        if nodeType == "Group" {
```

Comment:
Actually all these `if` conditions can be replaced if you will create map of factories: `[<nodeType>:(<dictionary>) -> <node>]`. In this case you don't need to add `init` to each class, but declare `factories` and then call `factories[nodeType]!(dictionary)`

Reply:
Then there will be another list of if-s - to add all the factories to the map. Can't see why it would be better